# Step-1 Data Exploration and Preprocessing/ EDA

In [16]:
import pandas as pd


df=pd.read_csv(r'C:\Users\SAM HP\Desktop\KHI AI backup (till 05-Feb-24)\Karachi AI 2\Data Analyst (CDA)\Virtual Internship Projects\Quick Model Insurance Expected Claim Amount for our Insurance Department Use Case\DataSet\insurance_data.csv')
df

index  PatientID   age  gender   bmi  bloodpressure diabetic  children  \
0         0          1  39.0    male  23.2             91      Yes         0   
1         1          2  24.0    male  30.1             87       No         0   
2         2          3   NaN    male  33.3             82      Yes         0   
3         3          4   NaN    male  33.7             80       No         0   
4         4          5   NaN    male  34.1            100       No         0   
...     ...        ...   ...     ...   ...            ...      ...       ...   
1335   1335       1336  44.0  female  35.5             88      Yes         0   
1336   1336       1337  59.0  female  38.1            120       No         1   
1337   1337       1338  30.0    male  34.5             91      Yes         3   
1338   1338       1339  37.0    male  30.4            106       No         0   
1339   1339       1340  30.0  female  47.4            101       No         0   

     smoker     region     claim  
0        No  southeast   1121.87  
1        No  southeast   1131.51  
2        No  southeast   1135.94  
3        No  northwest   1136.40  
4        No  northwest   1137.01  
...     ...        ...       ...  
1335    Yes  northwest  55135.40  
1336    Yes  northeast  58571.07  
1337    Yes  northwest  60021.40  
1338    Yes  southeast  62592.87  
1339    Yes  southeast  63770.43  

[1340 rows x 11 columns]

In [17]:
print("\nData types of columns:")
print(df.dtypes)


Data types of columns:
index              int64
PatientID          int64
age              float64
gender            object
bmi              float64
bloodpressure      int64
diabetic          object
children           int64
smoker            object
region            object
claim            float64
dtype: object


In [19]:
print("\nSummary statistics:")
df.describe()


Summary statistics:


index    PatientID          age          bmi  bloodpressure  \
count  1340.000000  1340.000000  1335.000000  1340.000000    1340.000000   
mean    669.500000   670.500000    38.078652    30.668955      94.157463   
std     386.968991   386.968991    11.102924     6.106735      11.434712   
min       0.000000     1.000000    18.000000    16.000000      80.000000   
25%     334.750000   335.750000    29.000000    26.275000      86.000000   
50%     669.500000   670.500000    38.000000    30.400000      92.000000   
75%    1004.250000  1005.250000    47.000000    34.700000      99.000000   
max    1339.000000  1340.000000    60.000000    53.100000     140.000000   

          children         claim  
count  1340.000000   1340.000000  
mean      1.093284  13252.745642  
std       1.205334  12109.609288  
min       0.000000   1121.870000  
25%       0.000000   4719.685000  
50%       1.000000   9369.615000  
75%       2.000000  16604.305000  
max       5.000000  63770.430000

In [20]:
print("\nMissing values:")
df.isnull().sum()


Missing values:


index            0
PatientID        0
age              5
gender           0
bmi              0
bloodpressure    0
diabetic         0
children         0
smoker           0
region           3
claim            0
dtype: int64

# Step-2 Feature Engineering

In [21]:
# Handle missing values
median_age = df['age'].median()
df['age'].fillna(median_age, inplace=True)

In [22]:
mode_region = df['region'].mode()[0]
df['region'].fillna(mode_region, inplace=True)

In [27]:
#Checking again out if we have still missing values
df.isnull().sum()

index            0
PatientID        0
age              0
gender           0
bmi              0
bloodpressure    0
diabetic         0
children         0
smoker           0
region           0
claim            0
dtype: int64

In [28]:
# One-hot encode categorical variables
df = pd.get_dummies(df, columns=['gender', 'diabetic', 'smoker'], drop_first=True)
df = pd.get_dummies(df, columns=['region'], drop_first=False)  # Preserve all unique values in 'region'

In [29]:
df

index  PatientID   age   bmi  bloodpressure  children     claim  \
0         0          1  39.0  23.2             91         0   1121.87   
1         1          2  24.0  30.1             87         0   1131.51   
2         2          3  38.0  33.3             82         0   1135.94   
3         3          4  38.0  33.7             80         0   1136.40   
4         4          5  38.0  34.1            100         0   1137.01   
...     ...        ...   ...   ...            ...       ...       ...   
1335   1335       1336  44.0  35.5             88         0  55135.40   
1336   1336       1337  59.0  38.1            120         1  58571.07   
1337   1337       1338  30.0  34.5             91         3  60021.40   
1338   1338       1339  37.0  30.4            106         0  62592.87   
1339   1339       1340  30.0  47.4            101         0  63770.43   

      gender_male  diabetic_Yes  smoker_Yes  region_northeast  \
0            True          True       False             False   
1            True         False       False             False   
2            True          True       False             False   
3            True         False       False             False   
4            True         False       False             False   
...           ...           ...         ...               ...   
1335        False          True        True             False   
1336        False         False        True              True   
1337         True          True        True             False   
1338         True         False        True             False   
1339        False         False        True             False   

      region_northwest  region_southeast  region_southwest  
0                False              True             False  
1                False              True             False  
2                False              True             False  
3                 True             False             False  
4                 True             False             False  
...                ...               ...               ...  
1335              True             False             False  
1336             False             False             False  
1337              True             False             False  
1338             False              True             False  
1339             False              True             False  

[1340 rows x 14 columns]

In [30]:
df.dtypes

index                 int64
PatientID             int64
age                 float64
bmi                 float64
bloodpressure         int64
children              int64
claim               float64
gender_male            bool
diabetic_Yes           bool
smoker_Yes             bool
region_northeast       bool
region_northwest       bool
region_southeast       bool
region_southwest       bool
dtype: object

# Step-3 Model Building

In [31]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from xgboost import XGBRegressor
from sklearn.svm import SVR
from sklearn.metrics import mean_absolute_error

# Split the dataset into features (X) and target variable (y)
X = df.drop(columns=['claim'])
y = df['claim']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize and train the models
linear_model = LinearRegression()
linear_model.fit(X_train, y_train)

rf_model = RandomForestRegressor()
rf_model.fit(X_train, y_train)

gb_model = GradientBoostingRegressor()
gb_model.fit(X_train, y_train)

xgb_model = XGBRegressor()
xgb_model.fit(X_train, y_train)

svr_model = SVR()
svr_model.fit(X_train, y_train)

# Make predictions on the test set
y_pred_linear = linear_model.predict(X_test)
y_pred_rf = rf_model.predict(X_test)
y_pred_gb = gb_model.predict(X_test)
y_pred_xgb = xgb_model.predict(X_test)
y_pred_svr = svr_model.predict(X_test)

# Evaluate the models
mae_linear = mean_absolute_error(y_test, y_pred_linear)
mae_rf = mean_absolute_error(y_test, y_pred_rf)
mae_gb = mean_absolute_error(y_test, y_pred_gb)
mae_xgb = mean_absolute_error(y_test, y_pred_xgb)
mae_svr = mean_absolute_error(y_test, y_pred_svr)

print(f'Linear Regression: Mean Absolute Error = {mae_linear}')
print(f'Random Forest Regressor: Mean Absolute Error = {mae_rf}')
print(f'Gradient Boosting Regressor: Mean Absolute Error = {mae_gb}')
print(f'XGBoost Regressor: Mean Absolute Error = {mae_xgb}')
print(f'Support Vector Regression: Mean Absolute Error = {mae_svr}')

#The Mean Absolute Error (MAE) value you obtained is a metric used to evaluate the performance of a regression model. It represents the average absolute difference between the predicted values and the actual values in the test set. 
#The MAE measures the average magnitude of errors in a set of predictions, without considering their direction (whether they are overestimations or underestimations).


Linear Regression: Mean Absolute Error = 3227.128371186802
Random Forest Regressor: Mean Absolute Error = 50.53432014925484
Gradient Boosting Regressor: Mean Absolute Error = 99.7098112487969
XGBoost Regressor: Mean Absolute Error = 138.41010824175018
Support Vector Regression: Mean Absolute Error = 8551.021831044982


In [32]:
models_data = {
    'liner': mae_linear,
    'rf': mae_rf,
    'gb': mae_gb,
    'xgb':mae_xgb,
    'svr':mae_svr
}

In [33]:
models_data

{'liner': 3227.128371186802,
 'rf': 50.53432014925484,
 'gb': 99.7098112487969,
 'xgb': 138.41010824175018,
 'svr': 8551.021831044982}

In [36]:
#new = pd.DataFrame.from_dict(final_data,orient='index')
Model_Results = pd.DataFrame(list(models_data.items()), columns=['Models', 'Values'])

In [37]:
Model_Results

Models       Values
0  liner  3227.128371
1     rf    50.534320
2     gb    99.709811
3    xgb   138.410108
4    svr  8551.021831

Considering that the Random Forest Regressor's Mean Absolute Error of 50.53432014925484 is the lowest of the models, it can be utilized as a pipeline for the deployment of streamlit-APP

In [38]:
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestRegressor

# Define the pipeline steps
pipeline = Pipeline([
    ('random_forest', RandomForestRegressor())
])

# Train the pipeline
pipeline.fit(X_train, y_train)

# Make predictions
y_pred_rf_pipeline = pipeline.predict(X_test)

# Evaluate the model
mae_rf_pipeline = mean_absolute_error(y_test, y_pred_rf_pipeline)

print(f'Random Forest Regressor (Pipeline): Mean Absolute Error = {mae_rf_pipeline}')

Random Forest Regressor (Pipeline): Mean Absolute Error = 50.9572608208972


In [39]:
import pickle
filename = 'pipeline.pkl'
pickle.dump(pipeline, open(filename,'wb'))

I have now developed a pipeline file that can be loaded into a streamlit application.